### ISSUE TITLE
R.flatten only accepts objects extending any[], but will flatten many other things when nested

### ISSUE THREAD

**Issue by aaronyo** [ramda/ramda#3182]
I got tripped up by ```flatten``` being a bit more aggressive than I thought it would be, in particular flattening HTMLFormElements.

Personally, i think the internal ```_isArrayLike``` is too aggressive.

But in the least, shouldn't ```flatten``` accept anything that passes _isArrayLike?

Looked at another way, shouldn't the following always hold true?

if: R.flatten([someVariable]) equals [1, 2]
then: someVariable is array like, and R.flatten(someVariable) also equals [1,2]

Presently, that's not the case, at least not in TypeScript.

**Comment by CrossEye** [comment]
> Personally, i think the internal `_isArrayLike` is too aggressive.

Do you have suggestions for a change to this?  I don't think there is a perfect answer that balances all the flexibility we might want with the simplicity we insist upon.  Note that the definition was specifically intended to capture things like `NodeList`s; it does feel as though `HTMLFormElement`s would naturally come along for the ride.  We might well have the balance wrong.  How would you change it?  

> However, in TypeScript, such objects may not be accepted as the argument to R.flatten, because R.flatten only accepts objects extending `any[]`, a subset of _isArrayLike.

I think this would have to be taken up with whoever maintains the Typescript typings you use.  The Ramda core team does not have a lot of TS expertise.

**Comment by aaronyo** [comment]
Thanks for the quick reply.

> > Personally, i think the internal `_isArrayLike` is too aggressive.
> 
> Do you have suggestions for a change to this? I don't think there is a perfect answer that balances all the flexibility we might want with the simplicity we insist upon. Note that the definition was specifically intended to capture things like `NodeList`s; it does feel as though `HTMLFormElement`s would naturally come along for the ride. We might well have the balance wrong. How would you change it?
> 

Good question, sorry about my vague critique, and thank you for the thoughtful reply. I'm sure others have thought about this a lot more than me.

TL;DR:
- Latest R version actually cleared up a big source of my concern
- I think _isArrayLike should be applied to R.flatten's argument, not just its argument's children. Otherwise, some objects will not be flattened when passed as the argument to R.flatten, but will be flattened when passed as a child of the arg.

After discovering my bug late last night, I looked at my local version of Ramda 0.27.0, and foolishly didn't think to look at 0.27.1 before posting. There's a relevant fix on 0.27.1:
https://github.com/ramda/ramda/commit/af3d323e3e6ec208799b7bd0736eca14d6849aff#diff-f6b10cd969d4e04bbbf1244edf131a806cb16e76cddd11f630133a22a4d88004
 
Now R.flatten should treat a zero length HTMLFormElement the same (flattening it) whether it is the argument or a child of the argument (not the case in 0.27.0).

Also the code doesn't seem to specifically target DOM nodes (and behave slightly differently with them) anymore. It seems the strategy is just to flatten anything that appears to have indexes and a length. Not sure how I feel about that definition of _isArrayLike, but I get it now, and am not sure what is best for R's users. (The fact that in JS land you'll find numeric indexes on objects that are not only lists seems to make this all a bit messy. I lean towards *not* flattening such objects, but I can imagine reasons to, as well. Not expert enough in relative concepts like Symbol.Iterator to really flesh out an alternative. Might dig into that a bit...)

That said, for some object types, in particular ones that have a length but then fail this criteria:
https://github.com/ramda/ramda/blob/fe4a05528f1b08cad45f61f0db34501c51cee650/source/internal/_isArrayLike.js#L30-L31
..those will be "flattened" when they are _children_ of an argument to R.flatten, but not when they _are_ the argument. You almost have to be trying to game R.flatten to trip this, but the fix is simple -- check the argument, not just its children, against _isArrayLike.

I guess the question is: do you agree with my expected behavior? Example:

expected a or b:

a)
```
> _.flatten([{0: 'a', length: 2}])
[ { '0': 'a', length: 2 } ]
> _.flatten({0: 'a', length: 2})
[ ]
```

b)
```
> _.flatten([{0: 'a', length: 2}])
[ 'a', undefined ]
> _.flatten({0: 'a', length: 2})
[ 'a', undefined ]
```

but actual is:
```
> _.flatten([{0: 'a', length: 2}])
[ { '0': 'a', length: 2 } ]
> _.flatten({0: 'a', length: 2})
[ 'a', undefined ]
```

> > However, in TypeScript, such objects may not be accepted as the argument to R.flatten, because R.flatten only accepts objects extending `any[]`, a subset of _isArrayLike.
> 
> I think this would have to be taken up with whoever maintains the Typescript typings you use. The Ramda core team does not have a lot of TS expertise.

Ah, that makes sense. I guess the DefinitivelyTyped repo is where I should have gone with that. Thank you.

**Comment by CrossEye** [comment]
> I think _isArrayLike should be applied to R.flatten's argument, not just its argument's children. Otherwise, some objects will not be flattened when passed as the argument to R.flatten, but will be flattened when passed as a child of the arg.

That makes perfect sense to me.  At first glance it seems exactly right.  I will think about it a little more, but in the meantime, are you interested in creating a PR for this?

As to handling HTML elements, the commit you pointed to was for #2836 and removed the very specific code, but we are still left with

```js
  if (x.length > 0) {
    return x.hasOwnProperty(0) && x.hasOwnProperty(x.length - 1);
  }
```

which I believe should capture most iterable HTML elements.

**Comment by aaronyo** [comment]
I have a busy couple of weeks coming up, but I would like to help out and create a PR. Should be a pretty quick change.

**Comment by CrossEye** [comment]
Pinging @aaronyo: are you still interested in creating a PR for this?
